cpp/dolfinx/nls/NewtonSolver.cpp

```diff
@@ -232,6 +228,9 @@ std::pair<int, bool> nls::petsc::NewtonSolver::solve(Vec x)
       // set.
       if (_iteration == 1)
       {
+         PetscReal _r = 0.0;
+         VecNorm(_dx, NORM_2, &_r);
+         _residual0 = _r;
         _residual = 1.0;
```

Comment:
Unrelated to the PR, but I think "residual" initialization `_residual = 1.0` for the incremental criterion should happen above, outside of the loop just as it does for residual criterion. In the current code, even if the first `||dx|| < atol` I would have to wait one additional iteration.

Besides the naming is not great and what we call "residual" for incremental criterion is actually increment norm...

Reply:
I think this piece of code could be completely refactored so that the convergence criterion can be switched out based on passing a different std::function?

Reply:
It can? You can do `set_convergence_check`?
https://github.com/search?q=repo%3AFEniCS%2Fdolfinx%20set_convergence_check&type=code
If you mean that we should have a function that includes the modification of the tolerances, I'm not sure that is worth-while.